File: edit_distance.py
class Solution:
    def minDistance(self,word1,word2):
        n=len(word1)
        m=len(word2)  
        dp=[[0 for x in range(m+1)] for y in range(n+1)]
        for j in range(m+1):
            dp[0][j]=j
        for i in range(n+1):
            dp[i][0]=i
        for i in range(0,n):
            for j in range(0,m):
                if word1[i]==word2[j]:
                    dp[i+1][j+1]=min(dp[i][j+1]+1,dp[i+1][j]+1,dp[i][j])
                else:
                    dp[i+1][j+1]=min(dp[i][j+1],dp[i+1][j],dp[i][j])+1
        return dp[n][m]
    def minDistance1(self,word1,word2):
        n=len(word1)
        m=len(word2) 
        dp=[[2**10 for x in range(m+1)] for y in range(n+1)]
        for j in range(m):
            dp[n][j]= m-j
        for i in range(n):
            dp[i][m] = n-i
        for i in range(n-1,-1,-1):
            for j in range(m-1,-1,-1):
                if word1[i]==word2[j]:
                    dp[i][j]=min(dp[i][j+1]+1,dp[i+1][j]+1,dp[i+1][j+1])
                else:
                    dp[i][j]=min(dp[i][j+1],dp[i+1][j],dp[i+1][j+1])+1
        return dp[0][0]
if __name__=='__main__':
    sol=Solution()
    word1='pneumonoultramicroscopicsilicovolcanoconiosis'
    word2='ultramicroscopically'
    print sol.minDistance(word1,word2)
    print sol.minDistance1(word1,word2)

File: Algorithm/CPP/edit_distance.cpp
# include <vector>
# include <algorithm>
# include <string>
# include <stack>
# include <algorithm>
# include <deque>
# include <iostream>
using namespace std;

class Solution {
public:
	int minDistance(string word1, string word2) {
		int m = word1.size();
		int n = word2.size();
		vector<vector<int>> dp(m+1, vector<int>(n+1));

		for (int i = 0; i < m+1; ++i)
		{
			dp[i][0] = i;
		}
		for (int j = 0; j < n+1; ++j)
		{
			dp[0][j] = j;
		}
		for (int i = 0; i < m; ++i)
			for (int j = 0; j < n; ++j)
			{
				if (word1[i] == word2[j])
				{
					dp[i+1][j+1] = min(min(dp[i][j+1], dp[i+1][j]) + 1, dp[i][j]);
				}
				else
				{
					dp[i+1][j+1] = min(dp[i][j], min(dp[i+1][j], dp[i][j+1])) + 1;
				}
			}
		return dp[m][n];

	}
};

int main()
{
	Solution sol;
	vector<int> nums = { 1, 1, 1, 3, 3, 2, 2, 2};
	//nums = { 1, 2 };
	int k = 8;
	//nums = { 3, 2, 1, 5, 6, 4 };
	string word1 = "vintner";
	string word2 = "writers";
	cout<<sol.minDistance(word1,word2)<< endl;
	//for (auto k :nums )cout << k << " ";
	system("pause");
	return 0;
}

